docs/rfcs/027-pageserver-migration.md

```diff
@@ -0,0 +1,454 @@
+  to download layers and catch up with WAL will take much longer.
+
+#### Cutover diagram
+
+```mermaid
+sequenceDiagram
+participant CP as Control plane
+participant A as Node A
+participant B as Node B
+participant E as Endpoint
+
+CP->>A: PUT Flush & go to AttachedStale
+note right of A: A continues to ingest WAL
+CP->>B: PUT AttachedMulti
+CP->>B: PUT Download image layers
+note right of B: B downloads from S3
+loop Poll until download complete
+CP->>B: GET download status
+end
+activate B
+note right of B: B ingests WAL
+loop Poll until catch up
+CP->>B: GET visible WAL
+CP->>A: GET visible WAL
+end
+deactivate B
+CP->>E: Configure to use Node B
+E->>B: Connect for reads
+CP->>B: PUT AttachedSingle
+CP->>A: PUT Secondary
+```
+
+#### Cutover from an unavailable pageserver
+
+This case is far simpler: we may skip straight to our intended
+end state.
+
+```mermaid
+sequenceDiagram
+participant A as Node A
+participant CP as Control plane
+participant B as Node B
+participant E as Endpoint
+
+note right of A: Node A offline
+activate A
+CP->>B: PUT AttachedSingle
+CP->>E: Configure to use Node B
+E->>B: Connect for reads
+deactivate A
+```
+
+## Implementation (detail)
+
+### Purpose of AttachedMulti, AttachedStale
+
+#### AttachedMulti
+
+Ordinarily, an attached pageserver whose generation is the latest may delete
+layers at will (e.g. during compaction). If a previous generation pageserver
+is also still attached, and in use by endpoints, then this layer deletion could
+lead to a loss of availability for the endpoint when reading from the previous
+generation pageserver.
+
+The _AttachedMulti_ state simply disables deletions. These will be enqueued
+until the control plane eventually transitions the node into AttachedSingle,
+which unblocks deletions.
+
+AttachedMulti is not required for data safety, only to preserve availability
+on pageservers running with stale generations.
+
+A node enters AttachedMulti only when explicitly asked to by the control plane.
+
+#### AttachedStale
+
+Currently, a pageserver with a stale generation number will continue to
+upload layers, but be prevented from completing deletions. This is safe, but inefficient: layers uploaded by this stale generation
+will not be read back by future generations of pageservers.
+
+The _AttachedStale_ state simply disables S3 uploads. The stale pageserver
+will continue to ingest the WAL and write layers to local disk, but not to
+do any uploads to S3.
+
+A node may enter AttachedStale in two ways:
+
+- Explicitly, when control plane calls into the node at the start of a migration.
+- Implicitly, when the node tries to validate some deletions and discovers
+  that its generation is tale.
+
+Over long periods of time, a node in AttachedStale will accumulate data
+on local disk, as it cannot evict any layers written since it entered the
+AttachStale state. We rely on the control plane to revert the location to
+Secondary or Detached at the end of a migration.
```

Comment:
This is insufficient.
Pageserver currently mustn't run out of disk space; the disk_usage_based_eviction prevents that.
But, it can't for these AttachedStale tenants, because their layers aren't remote.

So, effectively, AttachedStale prevents graceful degradation through trashing.

Not sure where / how we'd detect that situation in pageserver, but I fear we _must_ detect it if we want to have the AttachedStale state.

Alternatively, we could introduce AttachedStale later and not worry about it in this RFC?

Reply:
> Not sure where / how we'd detect that situation in pageserver, but I fear we must detect it if we want to have the AttachedStale state.

I agree with that statement: I had been somewhat implicitly relying on the control plane to notice a "too slow" migration and cancel it, to avoid this situation, but we can make this stronger by doing some enforcement on the pageserver side.  Things like:
- On AttachedStale, disable uploads _until_ we are asked to reclaim some disk space and would need to do uploads to achieve that.
- On AttachedStale, disable uploads for _up to_ a certain number of seconds (perhaps the default flush interval?), so that a very long running migration will revert to doing uploads.

> Alternatively, we could introduce AttachedStale later and not worry about it in this RFC?

I'd like to keep this in, so that it becomes part of the interface to the control plane and we don't have to revise that later.  The nice thing about AttachedStale is that it's totally advisory, so the pageserver can treat it as a no-op if we want to defer implementing a safer way to only stop doing uploads until disk pressure motivates us to do them.

Reply:
I've added a section `Disk pressure & AttachedStale` that addresses this.  Basically if we get evict_layers called on a Timeline for some  not-yet-uploaded layers, it will unblock uploads.

Reply:
Worst case if we have issues during implementation, we just make AttachedStale a no-op and tolerate the leaked objects: but I think it needs to go into the interface to the control plane to avoid disrupting the interface later as we refine the implementation.